Sources/Datadog/RUM/Instrumentation/Views/UIKit/UIKitHierarchyInspector.swift

```diff
@@ -1,66 +0,0 @@
-/*
- * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
- * This product includes software developed at Datadog (https://www.datadoghq.com/).
- * Copyright 2019-2020 Datadog, Inc.
- */
-
-import UIKit
-
-/// Inspects the view controllers hierarchy of the current window and finds
-/// the `UIViewController` which is currently displayed to the user.
-internal protocol UIKitHierarchyInspectorType {
-    /// The top-most `UIViewController` currently displayed to the user.
-    func topViewController() -> UIViewController?
-}
-
-internal struct UIKitHierarchyInspector: UIKitHierarchyInspectorType {
-    private let rootViewControllerProvider: () -> UIViewController?
-
-    init(
-        rootViewControllerProvider: @escaping () -> UIViewController? = {
-            let keyWindow = UIApplication.managedShared?.windows.first { $0.isKeyWindow }
-            return keyWindow?.rootViewController
-        }
-    ) {
-        self.rootViewControllerProvider = rootViewControllerProvider
-    }
-
-    // MARK: - UIKitHierarchyInspectorType
-
-    func topViewController() -> UIViewController? {
-        if let rootVC = rootViewControllerProvider(),
-           let highestPresentedVC = findHighestPresentedViewControllerInHierarchy(of: rootVC) {
-            return findTopViewControllerInHierarchy(of: highestPresentedVC)
-        }
-        return nil
-    }
-
-    // MARK: - Private
-
-    private func findHighestPresentedViewControllerInHierarchy(of viewController: UIViewController) -> UIViewController? {
-        if let presentedVC = viewController.presentedViewController {
-            return findHighestPresentedViewControllerInHierarchy(of: presentedVC)
-        }
-        return viewController
-    }
-
-    private func findTopViewControllerInHierarchy(of viewController: UIViewController) -> UIViewController? {
-        if let tabBarVC = viewController as? UITabBarController {
-            guard let selectedVC = tabBarVC.selectedViewController else {
-                return tabBarVC // when `UITabBarController` displays no view controllers
-            }
-            // When `UITabBarController` has selected VC, continue searching, as the selected VC
-            // may be containing another hierarchy (e.g. `UINavigationController`).
-            return findTopViewControllerInHierarchy(of: selectedVC)
-        } else if let navigationVC = viewController as? UINavigationController {
-            guard let topVC = navigationVC.topViewController else {
-                return navigationVC // when `UINavigationController` displays no view controllers
-            }
-            // When `UINavigationController` has top VC, continue searching, as the top VC
-            // may be containing another hierarchy (e.g. `UITabBarController`).
-            return findTopViewControllerInHierarchy(of: topVC)
-        } else {
-            return viewController
-        }
-    }
```

Comment:
All these are smoothly supported with the new approach of tracking the `stack`? Or do you think there are still some edge cases we need to look closer at?

Reply:
`UITabBarController` and `UINavigationController` are both supported by the stack, it send events in the same manner. I did some manual monkey testing to validate it, and integration tests are both green. @ncreated do you see any other checks I should run? (don't know if e2e tests applies here).

I can think of one edge case which I need to verify: When presenting multiple modal page sheet and one of them is rejected by the predicate. Then the back navigation can be inconsistent:
1. start view 1
2. reject view 2
3. start view 3
4. Dismiss view 3 -> will restart view 1 ❌
 > This is due to the fact that the vc below a page sheet does not disappear, in this case view 1 will still be in the stack

Reply:
I see 👍. Yes, this new approach will require quite manual testing, as we're entirely changing the model. The initial reason for introducing `UIKitHierarchyInspector` was to deal with the complexity of back and forth transitions between `UIViewControllers`, notably modal view controllers. Test fixtures were added to `Example` app in #271 and the implementation of `UIKitHierarchyInspector` was done in #278.

Except having integration tests pass, we must do manual tests for the new implementation as well and I'll be happy to help on that later 🙌. Our integration tests do not cover all complexity (that would be tedious), and there might be still scenarios worth manual testing, e.g.:
* changing the entire navigation tree, by `window.rootViewController = ...`;
* dismissing first VC in the modal hierarchy of `N` VCs (it should dismiss the entire hierarchy);
* changing `navigationController.viewControllers` stack programmatically;
* jumping by few steps at once in the VCs hierarchy (e.g. imagine app receiving deeplink and handling it by pushing one VC to the stack + opening modal with another navigation VC).

I also remember that our `UIKitHierarchyInspector` was not well compatible with `UISearchController` and handling it required additional logic in the "views predicate". This compatibility could be best tested when dogfooding a branch in our Datadog iOS App, as it uses `UISearchController` in many places. We should have a look into this at the end, before final merge.

Reply:
I have performed and validate the following manual testing by modifying the Example scenarios:
- changing the entire navigation tree, by `window.rootViewController = ...` ✅ 
- dismissing first VC in the modal hierarchy of `N` VCs, it should dismiss the entire hierarchy ✅ 
- changing `navigationController.viewControllers` stack programmatically ✅ 
- jumping by few steps at once in the VCs hierarchy (e.g. imagine app receiving deeplink and handling it by pushing one VC to the stack + opening modal with another navigation VC). ✅ 

I still need to test it with a `UISearchController`!

Reply:
> I still need to test it with a UISearchController!

Keep in mind we don't yet support `UISearchController` very well - it still requires manual tweaking in the implementation of `predicate` to support search bar transitions to embedded VCs. I think it will be enough to just do regression test on `UISearchController` and check if it gets recognised by `UIKit` instrumentation as a whole. Solving the problem with search bar transitions to embedded VCs is beyond the scope of this PR.